worker/remoterelations/remoterelations.go

```diff
@@ -388,86 +369,117 @@ func (w *remoteApplicationWorker) relationChanged(
-			w.relationInfo.remoteEndpointName = remoteRelation.RemoteEndpointName
-			var err error
-			remoteRelationId, err = w.registerRemoteRelation(relationTag, publisher)
-			if err != nil {
-				return errors.Trace(err)
-			}
-		}
+	}
+	if remoteRelation.Life == params.Dead {
+		// We haven't started the relation unit watcher so just exit.
+		return nil
 	}
 
-	// Start a watcher to track changes to the local units in the
-	// relation, and a worker to process those changes.
-	if remoteRelation.Life != params.Dead {
-		localRelationUnitsWatcher, err := w.facade.WatchLocalRelationUnits(key)
+	var remoteRelationId params.RemoteEntityId
+	if w.registered {
+		// We are on the offering side and the relation has been registered,
+		// so look up the token to use when communicating status.
+		token, err := w.facade.GetToken(w.remoteModelUUID, relationTag)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		relationUnitsWatcher, err := newRelationUnitsWatcher(
-			relationTag,
-			w.relationInfo.applicationId,
-			remoteRelationId,
-			localRelationUnitsWatcher,
-			w.relationChanges,
-			w.facade,
-			w.exportMutex,
-		)
+		remoteRelationId = params.RemoteEntityId{ModelUUID: w.remoteModelUUID, Token: token}
+		// Look up the exported token of the local application in the relation.
+		// The export was done when the relation was registered.
+		token, err = w.facade.GetToken(w.localModelUUID, names.NewApplicationTag(remoteRelation.LocalApplicationName))
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if err := w.catacomb.Add(relationUnitsWatcher); err != nil {
+		w.relationInfo.applicationId = params.RemoteEntityId{ModelUUID: w.localModelUUID, Token: token}
+	} else {
+		// We have not seen the relation before, make
+		// sure it is registered on the offering side.
+		w.relationInfo.localEndpoint = remoteRelation.LocalEndpoint
+		w.relationInfo.remoteEndpointName = remoteRelation.RemoteEndpointName
+		var err error
+		applicationTag := names.NewApplicationTag(remoteRelation.LocalApplicationName)
+		w.relationInfo.applicationId, remoteRelationId, err = w.registerRemoteRelation(applicationTag, relationTag, publisher)
+		if err != nil {
 			return errors.Trace(err)
 		}
-		r := &relation{}
-		r.RelationId = remoteRelation.Id
-		r.Life = remoteRelation.Life
-		r.ruw = relationUnitsWatcher
-		relations[key] = r
 	}
+
+	// Start a watcher to track changes to the local units in the
+	// relation, and a worker to process those changes.
+	localRelationUnitsWatcher, err := w.facade.WatchLocalRelationUnits(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	relationUnitsWatcher, err := newRelationUnitsWatcher(
+		relationTag,
+		w.relationInfo.applicationId,
+		remoteRelationId,
+		localRelationUnitsWatcher,
+		w.relationChanges,
+		w.facade,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(relationUnitsWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	r := &relation{}
```

Comment:
It's a bit messier than that, but changed

	relations[key] = &relation{
		RemoteRelationChange: params.RemoteRelationChange{
			RelationId: remoteRelation.Id,
			Life:       remoteRelation.Life,
		},
		ruw: relationUnitsWatcher,
	}